controller/lib/hal/hal_stm32.cpp

```diff
@@ -0,0 +1,856 @@
+// but the HAL interface currently isn't set up to run that way.  That's an improvement
+// that can be made later
+int HalApi::analogRead(AnalogPin pin){
+
+   int channel = 0;
+   switch( pin )
+   {
+      case AnalogPin::PATIENT_PRESSURE:
+         channel = 6;
+         break;
+
+      case AnalogPin::INFLOW_PRESSURE_DIFF:
+         channel = 9;
+         break;
+
+      case AnalogPin::OUTFLOW_PRESSURE_DIFF:
+         channel = 15;
+         break;
+
+      default:
+         return 0;
+   }
+
+   ADC_Regs *adc = (ADC_Regs *)ADC_BASE;
+   adc->adc[0].seq[0] = channel << 6;
+
+   // Clear the EOC flag
+   adc->adc[0].stat = 4;
+
+   // Start the conversion
+   adc->adc[0].ctrl |= 4;
+
+   // Wait for the end of the conversion
+   while( !(adc->adc[0].stat & 4) ){}
+
+   // Return the result
+   return adc->adc[0].data;
+}
+
+/******************************************************************
+ * PWM outputs
+ *
+ * The following four outputs could be driven
+ * as PWM outputs:
+ *
+ * PA8  - Timer 1 Channel 1 - heater control
+ * PA11 - Timer 1 Channel 4 - solenoid
+ * PB3  - Timer 2 Channel 2 - blower control
+ * PB4  - Timer 3 Channel 1 - buzzer
+ *
+ * For now I'll just set up the blower since that's the only
+ * one called out in the HAL
+ *****************************************************************/
+static void InitPwmOut( void )
+{
+   // The PWM frequency isn't mentioned anywhere that I can find, so
+   // I'm just picking a reasonable number.  This can be refined later
+   const int pwmFreqHz  = 20000;
+
+   EnableClock( TIMER2_BASE );
+
+   // Connect PB3 to timer 2
+   GPIO_PinAltFunc( GPIO_B_BASE, 3, 1 );
+
+   TimerRegs *tmr = (TimerRegs *)TIMER2_BASE;
+
+   // Set the frequency
+   tmr->reload = (CPU_FREQ / pwmFreqHz)-1;
+
+   // Configure channel 2 in PWM output mode
+   tmr->ccMode[0] = 0x6800;
+
+   tmr->ccEnable = 0x10;
+
+   // Start with 0% duty cycle
+   tmr->compare[1] = 0;
+
+   // Load the shadow registers
+   tmr->event = 1;
+
+   // Start the counter
+   tmr->ctrl[0] = 0x81;
+}
+
+// Set the PWM period.  Currently the value is 0 to 255 which is
+// a shame.  We should change this to a float, say with a value 
+// of 0.0 to 1.0 to make better use of our resolution
+void HalApi::analogWrite(PwmPin pin, int value) {
+
+   // Convert the value to a float
+   float duty =  value * (1.0/255);
+
+   TimerRegs *tmr;
+   int chan;
+   switch( pin )
+   {
+      case PwmPin::BLOWER:
+         tmr = (TimerRegs *)TIMER2_BASE;
+         chan = 1;
+         break;
+
+      default:
+         return;
+   }
+
+   tmr->compare[chan] = tmr->reload * duty;
+}
+
+/******************************************************************
+ * Serial port to GUI
+ *****************************************************************/
+
+// This class is a generic circular buffer for byte sized data.
+// It could probably go in it's own header outside the HAL, we 
+// would just need to add interrupt suspend/restore to the HAL
```

Comment:
I think this is warranted. I definitely would not trust myself to write a circular buffer without unit tests correctly, and this will have to be extracted in order to write unit tests for it.

Reply:
I agree.  Just wanted to limit the scope of the PR.  Once this is merged I'll create a new PR pulling this out and adding unit tests for it.